`apmperf`: panic slice out of bounds

**Describe the bug**

Encountered this panic when running apmsoak in the tilt/AWS setup

```
{"log.level":"error","@timestamp":"2024-03-19T14:54:09.768Z","log.logger":"soaktest","log.origin":{"function":"github.com/elastic/apm-perf/internal/loadgen/eventhandler.logResponseOutcome","file.name":"eventhandler/apm.go","file.line":90},"message":"request failed","status_code":401,"response":"{\"message\":\"Unauthorized\"}\n","ecs.version":"1.6.0"}
panic: runtime error: slice bounds out of range [1267810:759964]

goroutine 21 [running]:
bytes.(*Buffer).Bytes(...)
	/usr/local/go/src/bytes/buffer.go:54
net/http.NewRequestWithContext({0x26893d8?, 0x40001440f0}, {0x43673f?, 0x0?}, {0x400012a1b0?, 0x30?}, {0x2685380, 0x40000b4040?})
	/usr/local/go/src/net/http/request.go:876 +0x670
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*APMTransport).SendEvents(0x400011ca80, {0x26893d8?, 0x40001440f0?}, {0x2685380, 0x40000b4040}, 0x0?)
	/opt/apm-perf/internal/loadgen/eventhandler/apm.go:50 +0x60
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*Handler).sendBatch(0x400014a0c0, {0x26893d8, 0x40001440f0}, 0x40000a5b60, {{0x4001ab2240, 0x23b, 0x23b}, {0x4004e00000, 0xea9, 0x1300}}, ...)
	/opt/apm-perf/internal/loadgen/eventhandler/handler.go:358 +0x450
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*Handler).sendBatches(0x400014a0c0, {0x26893d8, 0x40001440f0}, 0x40000a5b60)
	/opt/apm-perf/internal/loadgen/eventhandler/handler.go:307 +0x390
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*Handler).SendBatchesInLoop(0x400014a0c0, {0x26893d8, 0x40001440f0})
	/opt/apm-perf/internal/loadgen/eventhandler/handler.go:269 +0x74
github.com/elastic/apm-perf/internal/soaktest.runAgent({0x26893d8, 0x40001440f0}, 0x4000126480, {{0x400013c140, 0x20}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
	/opt/apm-perf/internal/soaktest/runner.go:127 +0x204
github.com/elastic/apm-perf/internal/soaktest.(*Runner).Run.func1()
	/opt/apm-perf/internal/soaktest/runner.go:103 +0x128
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/go/pkg/mod/golang.org/x/sync@v0.6.0/errgroup/errgroup.go:78 +0x5c
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/pkg/mod/golang.org/x/sync@v0.6.0/errgroup/errgroup.go:75 +0x9c
```

Another one found in a more recent version

```
panic: runtime error: slice bounds out of range [923765:757053]

goroutine 54 [running]:
bytes.(*Buffer).Bytes(...)
	/usr/local/go/src/bytes/buffer.go:54
net/http.NewRequestWithContext({0x2c52410?, 0x400003c080}, {0x97c3f3?, 0x0?}, {0x400003a930?, 0x30?}, {0x2c474a0, 0x4000332040?})
	/usr/local/go/src/net/http/request.go:876 +0x670
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*APMTransport).SendEvents(0x4000116270, {0x2c52410?, 0x400003c080?}, {0x2c474a0, 0x4000332040}, 0x0?)
	/opt/apm-perf/internal/loadgen/eventhandler/apm.go:50 +0x60
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*Handler).sendBatch(0x4000190000, {0x2c52410, 0x400003c080}, 0x4000355ef0, {{0x4003ed4000, 0x23b, 0x23b}, {0x4004e80000, 0xea9, 0x1300}}, ...)
	/opt/apm-perf/internal/loadgen/eventhandler/handler.go:358 +0x450
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*Handler).sendBatches(0x4000190000, {0x2c52410, 0x400003c080}, 0x4000355ef0)
	/opt/apm-perf/internal/loadgen/eventhandler/handler.go:307 +0x390
github.com/elastic/apm-perf/internal/loadgen/eventhandler.(*Handler).SendBatches(0x4000190000, {0x2c52410, 0x400003c080})
	/opt/apm-perf/internal/loadgen/eventhandler/handler.go:288 +0x58
main.benchmarkAgent.func1(0x400019c160)
	/opt/apm-perf/cmd/apmbench/bench.go:152 +0xbc
testing.(*B).RunParallel.func1()
	/usr/local/go/src/testing/benchmark.go:815 +0xbc
created by testing.(*B).RunParallel
	/usr/local/go/src/testing/benchmark.go:808 +0xf0

```

--- a/internal/loadgen/eventhandler/apm-writer.go
+++ b/internal/loadgen/eventhandler/apm-writer.go
@@ -11,9 +11,9 @@ import (
 	"github.com/tidwall/gjson"
 )
 
-// writeAPMEvents writes to buffers in pooledWriter JSON formatted events that can be replayed.
+// writeAPMEvents writes to buffers in eventWriter JSON formatted events that can be replayed.
 // Implements EventWriter interface.
-func writeAPMEvents(config Config, minTimestamp time.Time, w *pooledWriter, b batch, baseTimestamp time.Time, randomBits uint64) error {
+func writeAPMEvents(config Config, minTimestamp time.Time, w *eventWriter, b batch, baseTimestamp time.Time, randomBits uint64) error {
 	rewriteAny := config.RewriteTimestamps ||
 		config.RewriteIDs ||
 		config.RewriteServiceNames ||

--- a/internal/loadgen/eventhandler/event.go
+++ b/internal/loadgen/eventhandler/event.go
@@ -12,4 +12,4 @@ type EventCollector interface {
 	Process([]byte) event
 }
 
-type EventWriter func(config Config, minTimestamp time.Time, w *pooledWriter, b batch, baseTimestamp time.Time, randomBits uint64) error
+type EventWriter func(config Config, minTimestamp time.Time, w *eventWriter, b batch, baseTimestamp time.Time, randomBits uint64) error

--- a/internal/loadgen/eventhandler/handler.go
+++ b/internal/loadgen/eventhandler/handler.go
@@ -56,10 +56,9 @@ type event struct {
 //
 // It is safe to make concurrent calls to Handler methods.
 type Handler struct {
-	mu         sync.Mutex // guards config.Rand
-	config     Config
-	writerPool sync.Pool
-	logger     *zap.Logger
+	mu     sync.Mutex // guards config.Rand
+	config Config
+	logger *zap.Logger
 
 	batches      []batch
 	minTimestamp time.Time // across all batches
@@ -181,17 +180,6 @@ func New(logger *zap.Logger, config Config, ec EventCollector) (*Handler, error)
 	h := Handler{
 		logger: logger.Named("handler"),
 		config: config,
-		writerPool: sync.Pool{
-			New: func() any {
-				pw := &pooledWriter{}
-				zw, err := zlib.NewWriterLevel(&pw.buf, zlib.BestSpeed)
-				if err != nil {
-					logger.Panic(err.Error())
-				}
-				pw.Writer = zw
-				return pw
-			},
-		},
 	}
 
 	matches, err := fs.Glob(config.Storage, config.Path)
@@ -296,11 +284,6 @@ func (h *Handler) sendBatches(ctx context.Context, s *state) (int, error) {
 	h.mu.Unlock()
 	h.logger.Debug("got random bits for batch")
 
-	s.w = h.writerPool.Get().(*pooledWriter)
-	defer h.writerPool.Put(s.w)
-	defer s.w.Reset()
-	h.logger.Debug("got writer from pool")
-
 	baseTimestamp := time.Now().UTC()
 	h.logger.Debug("calculated base timestamp", zap.String("timestamp", baseTimestamp.String()))
 	for _, batch := range h.batches {
@@ -341,8 +324,10 @@ func (h *Handler) sendBatch(
 				n = capacity
 			}
 		}
+
+		writer := newEventWriter()
 		if err := h.config.Writer(h.config, h.minTimestamp,
-			s.w, batch{
+			writer, batch{
 				metadata: b.metadata,
 				events:   events[:n],
 			}, baseTimestamp, randomBits); err != nil {
@@ -350,24 +335,26 @@ func (h *Handler) sendBatch(
 		}
 		h.logger.Debug("wrote events to buffer")
 
-		if err := s.w.Close(); err != nil {
+		if err := writer.Close(); err != nil {
 			return err
 		}
 		h.logger.Debug("closed writer")
 
-		if err := h.config.Transport.SendEvents(ctx, &s.w.buf, h.config.IgnoreErrors); err != nil {
+		// Do not reuse `writer`: in error cases, SendEvents may return while the request body is
+		// still being transmitted by the HTTP library. Reusing `writer` could cause a panic due to
+		// concurrent reads & writes.
+		if err := h.config.Transport.SendEvents(ctx, &writer.buf, h.config.IgnoreErrors); err != nil {
 			return err
 		}
 		h.logger.Debug("sent events through transport")
 
-		s.w.Reset()
 		s.sent += n
 		events = events[n:]
 	}
 	return nil
 }
 
-func rewriteJSONObject(w *pooledWriter, object gjson.Result, f func(key, value gjson.Result) bool) {
+func rewriteJSONObject(w *eventWriter, object gjson.Result, f func(key, value gjson.Result) bool) {
 	w.rewriteBuf.RawByte('{')
 	first := true
 	object.ForEach(func(key, value gjson.Result) bool {
@@ -462,21 +449,26 @@ func randomizeASCII(out *bytes.Buffer, in string, randomBits uint64) {
 }
 
 type state struct {
-	w     *pooledWriter
 	burst int
 	sent  int
 }
 
-type pooledWriter struct {
+type eventWriter struct {
 	rewriteBuf fastjson.Writer
 	idBuf      bytes.Buffer
 	buf        bytes.Buffer
 	*zlib.Writer
 }
 
-func (pw *pooledWriter) Reset() {
-	pw.rewriteBuf.Reset()
-	pw.idBuf.Reset()
-	pw.buf.Reset()
-	pw.Writer.Reset(&pw.buf)
+func newEventWriter() *eventWriter {
+	w := &eventWriter{}
+	// Preallocate to minimize memory copies.
+	w.buf.Grow(2 * 1024 * 1024)
+	zw, err := zlib.NewWriterLevel(&w.buf, zlib.BestSpeed)
+	if err != nil {
+		// Happens only when compression level doesn't exist
+		panic(err.Error())
+	}
+	w.Writer = zw
+	return w
 }

